### PR TITLE
Mttl-470 Asset Search Endpoint

### PR DESCRIPTION
**Pull request by evak6979**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/MTTL-470

## Describe this PR

Search endpoint for an Asset

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: HousingSearchApi/V1/Gateways/Models/QueryableAssetAddress.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        [Text(Name = "addressLine3")]
+        public string AddressLine3 { get; set; }
+
+        [Text(Name = "addressLine4")]
+        public string AddressLine4 { get; set; }
+
+        [Text(Name = "postCode")]
+        public string PostCode { get; set; }
+
+        [Text(Name = "postCode")]
```
Comment (humulla) [inline HousingSearchApi/V1/Gateways/Models/QueryableAssetAddress.cs:25]: Would we want the name different here so it not the same as the postCode identified above

**Review by humulla — COMMENTED** [review]
Do we not have any unit tests for the asset endpoint?

File: HousingSearchApi/V1/Helper/ModelValidationHelper.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            err.FieldName = "Insufficient characters";
+            errors.Add(err);
+
+            return new BadRequestObjectResult(new ErrorResponse(errors)
+            {
+                StatusCode = 400
+            });
+        }
+    }
+}
```
Comment (mikemoody-amido) [inline HousingSearchApi/V1/Helper/ModelValidationHelper.cs:25]: Not sure on the need for this. 
If the validation fails in the validator then the controller is not even used as MVC will return a 400 before we get here.

File: HousingSearchApi.Tests/V1/E2ETests/Fixtures/TenureFixture.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                var tenureSettingsDoc = File.ReadAllTextAsync("./data/elasticsearch/tenureIndex.json").Result;
+                ElasticSearchClient.LowLevel.Indices.CreateAsync<BytesResponse>(INDEX, tenureSettingsDoc)
+                    .ConfigureAwait(true);
+
+                var tenures = CreateTenureData();
+                var awaitable = ElasticSearchClient.IndexManyAsync(tenures, INDEX).ConfigureAwait(true);
+
+                while (!awaitable.GetAwaiter().IsCompleted)
+                {
+
```
Comment (mikemoody-amido) [inline HousingSearchApi.Tests/V1/E2ETests/Fixtures/TenureFixture.cs:39]: Timeout?

File: HousingSearchApi.Tests/V1/E2ETests/Fixtures/PersonsFixture.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                var personSettingsDoc = File.ReadAllTextAsync("./data/elasticsearch/personIndex.json").Result;
+                ElasticSearchClient.LowLevel.Indices.CreateAsync<BytesResponse>(INDEX, personSettingsDoc)
+                    .ConfigureAwait(true);
+
+                var persons = CreatePersonData();
+                var awaitable = ElasticSearchClient.IndexManyAsync(persons, INDEX).ConfigureAwait(true);
+
+                while (!awaitable.GetAwaiter().IsCompleted)
+                {
+
```
Comment (mikemoody-amido) [inline HousingSearchApi.Tests/V1/E2ETests/Fixtures/PersonsFixture.cs:40]: Timeout?

File: HousingSearchApi.Tests/V1/E2ETests/Fixtures/AssetFixture.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                var assetSettingsDoc = File.ReadAllTextAsync("./data/elasticsearch/assetIndex.json").Result;
+                ElasticSearchClient.LowLevel.Indices.CreateAsync<BytesResponse>(INDEX, assetSettingsDoc)
+                    .ConfigureAwait(true);
+
+                var assets = CreateAssetData();
+                var awaitable = ElasticSearchClient.IndexManyAsync(assets, INDEX).ConfigureAwait(true);
+
+                while (!awaitable.GetAwaiter().IsCompleted)
+                {
+
```
Comment (mikemoody-amido) [inline HousingSearchApi.Tests/V1/E2ETests/Fixtures/AssetFixture.cs:39]: Timeout?

File: HousingSearchApi/V1/Boundary/Requests/Validation/GetPersonListRequestValidator.cs [hunk 6]
```diff
@@ -3,7 +3,7 @@
 
 namespace HousingSearchApi.V1.Boundary.Requests.Validation
 {
-    public class GetPersonListRequestValidator : AbstractValidator<GetPersonListRequest>
+    public class GetPersonListRequestValidator : AbstractValidator<HousingSearchRequest>
```
[the remark this answers is not in the document]
Comment (evak6979) [reply HousingSearchApi/V1/Boundary/Requests/Validation/GetPersonListRequestValidator.cs:6]: I can just rename it if you want.

File: HousingSearchApi.Tests/V1/E2ETests/Fixtures/TenureFixture.cs (hunk 3 above)
Comment (evak6979) [reply HousingSearchApi.Tests/V1/E2ETests/Fixtures/TenureFixture.cs:39]: I tried to make it work like so, and it still failed. I have -no- clue what's with ES indexing and waiting on it...

File: HousingSearchApi/V1/Domain/Tenure/TenuredAsset.cs [hunk 7]
```diff
@@ -14,12 +14,17 @@ public static TenuredAsset Create(QueryableTenuredAsset tenuredAsset)
...
         private TenuredAsset(QueryableTenuredAsset tenuredAsset)
         {
-            FullAddress = tenuredAsset.FullAddress;
-            Uprn = tenuredAsset.Uprn;
-            Id = tenuredAsset.Id;
-            Type = tenuredAsset.Type;
+            FullAddress = tenuredAsset?.FullAddress;
+            Uprn = tenuredAsset?.Uprn;
+            Id = tenuredAsset?.Id;
+            Type = tenuredAsset?.Type;
```
[the remark this answers is not in the document]
Comment (evak6979) [reply HousingSearchApi/V1/Domain/Tenure/TenuredAsset.cs:27]: It broke a test of mine hence why I thought to double check for nulls.

**Review by mikemoody-amido — APPROVED** [review]

**Pull request merged**